test/types.spec.ts

```diff
@@ -156,4 +160,79 @@ describe('http-proxy-middleware TypeScript Types', () => {
       });
     });
   });
+
+  describe('request response inference', () => {
+    interface FooBarRequest extends http.IncomingMessage {
+      foo: string;
+    }
+    interface FooBarResponse extends http.ServerResponse {
+      bar: number;
+    }
+    const fooBarUse = (handler: (request: FooBarRequest, response: FooBarResponse) => void) => {};
+
+    fooBarUse(createProxyMiddleware((_, request) => request.foo && true));
+    fooBarUse(
+      createProxyMiddleware({
+        onError: (_, request, response) => {
+          request.foo;
+          response.bar;
+          // @ts-expect-error
+          request.params;
+          // @ts-expect-error
+          response.json;
+        },
+      })
+    );
+  });
+
+  describe('works for third-party libraries', () => {
+    express().use(
+      '/proxy',
+      createProxyMiddleware({
+        onError: (_, request, response) => {
+          request.params;
+          response.json;
+          // @ts-expect-error
+          request.lol;
+          // @ts-expect-error
+          response.lol;
+        },
+      })
+    );
+
+    connect().use(
+      '/proxy',
+      createProxyMiddleware({
+        onError: (_, request, response) => {
+          // todo, non trivial fix @ts-expect-error
```

Comment:
This can be fixed by sending a PR to `@types/connect` to have the following change over [here](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/connect/index.d.ts#L56-L57)
```diff
-        use(fn: HandleFunction): Server;
+        use(fn: SimpleHandleFunction): Server;
+        use(fn: NextHandleFunction): Server;
+        use(fn: ErrorHandleFunction): Server;
-        use(route: string, fn: HandleFunction): Server;
+        use(route: string, fn: SimpleHandleFunction): Server;
+        use(route: string, fn: NextHandleFunction): Server;
+        use(route: string, fn: ErrorHandleFunction): Server;
```
...this would make the inference work. There's no other way to fix it on our end.

Now the question is: Are the improvements in this PR worth having http-proxy-middleware to not work with connect properly? I'm okay with all answers although I would push "Yes" because as http-proxy-middleware should not be responsible for type gotchas of connect, nor should it's users be affected.